lms/djangoapps/commerce/api/v1/tests.py

```diff
@@ -0,0 +1,155 @@
+""" API v1 tests. """
+import json
+
+import ddt
+from django.conf import settings
+from django.contrib.auth.models import Permission
+from django.core.urlresolvers import reverse
+from django.test.utils import override_settings
+from django.utils import timezone
+from xmodule.modulestore.tests.django_utils import ModuleStoreTestCase
+from xmodule.modulestore.tests.factories import CourseFactory
+
+from commerce.api.v1.serializers import DATETIME_FORMAT
+from course_modes.models import CourseMode
+from student.tests.factories import UserFactory
+
+PASSWORD = 'test'
+JSON_CONTENT_TYPE = 'application/json'
+
+
+class CourseApiViewTestMixin(object):
+    """ Mixin for CourseApi views.
+
+    Automatically creates a course and CourseMode.
+    """
+
+    def setUp(self):
+        super(CourseApiViewTestMixin, self).setUp()
+        self.course = CourseFactory.create()
+        self.course_mode = CourseMode.objects.create(course_id=self.course.id, mode_slug=u'verified', min_price=100,
+                                                     currency=u'USD', sku=u'ABC123', expiration_datetime=timezone.now())
+
+    @staticmethod
+    def _serialize_course_mode(course_mode):
+        """ Serialize a CourseMode to a dict. """
+
+        expires = course_mode.expiration_datetime.strftime(DATETIME_FORMAT) if course_mode.expiration_datetime else None
+        return {
+            u'name': course_mode.mode_slug,
+            u'currency': course_mode.currency,
+            u'price': course_mode.min_price,
+            u'sku': course_mode.sku,
+            u'expires': expires
+        }
+
+
+class CourseListViewTests(CourseApiViewTestMixin, ModuleStoreTestCase):
+    """ Tests for CourseListView. """
+    path = reverse('commerce:api:v1:courses:list')
+
+    def test_authentication_required(self):
+        """ Verify only authenticated users can access the view. """
+        self.client.logout()
+        response = self.client.get(self.path, content_type=JSON_CONTENT_TYPE)
+        self.assertEqual(response.status_code, 401)
+
+    def test_list(self):
+        """ Verify the view lists the available courses and modes. """
+        user = UserFactory.create()
+        self.client.login(username=user.username, password=PASSWORD)
+        response = self.client.get(self.path, content_type=JSON_CONTENT_TYPE)
+
+        self.assertEqual(response.status_code, 200)
+        actual = json.loads(response.content)
+        expected = [
+            {
+                u'id': unicode(self.course.id),
+                u'modes': [self._serialize_course_mode(self.course_mode)]
+            }
+        ]
+        self.assertListEqual(actual, expected)
+
+
+@ddt.ddt
+class CourseRetrieveUpdateViewTests(CourseApiViewTestMixin, ModuleStoreTestCase):
```

Comment:
Ah. `test_retrieve_invalid_course` added to handle the GET. PUT will not return a 404 for now, as it does not utilize that block of code.